docs/available-filters/envelope-encryption/envelope-encryption.adoc

```diff
@@ -0,0 +1,210 @@
+
+=== Key rotation
+
+Key rotation is the practice of periodically replacing cryptographic keys with new ones.  Using key rotation is
+considered cryptographic best-practice.
+
+The filter allows for the rotation of KEKs within the KMS. When a KEK is rotated, the new key material will be applied
+to newly produced records. Existing records (which were encrypted with older versions of the KEK) remain decryptable
+as long as the previous KEK version remains present in the KMS.
+
+WARNING: If the previous KEK version is removed from the KMS, the records encrypted with that key version will become
+un-consumable (that is, the fetch will fail). In this case, the consumer offset must be advanced beyond those records.
+
+=== What exactly gets encrypted?
+
+The filter currently encrypts only *record values*.  Record keys, headers, timestamps are not encrypted.
+
+Null record values produced are passed through unencrypted.  This serves to allow tombstoning.
+
+The filter supports use-cases where some Kafka topics are configured for encryption while others are left to be
+unencrypted.  Transactions producing to both encrypted and unencrypted topics are unsupported?
+
+== How to use the filter
+
+There are three steps to using the filter.
+
+1. Setting up the KMS.
+2. Configuring the filter within Kroxylicious.
+3. Establishing the encryption key(s) within the KMS that will be used to encrypt the topics.
+
+These steps are described in the next sections.
+
+=== Setting up the KMS
+
+In order to set up the KMS provider ready to use with the filter, follow these KMS provider specific steps.
+
+include::hashicorp-vault/setup.adoc[leveloffset=3]
+
+=== Filter Configuration
+
+The filter is configured as part of the filter chain in the following way:
+
+[source, yaml]
+----
+filters:
+- type: EnvelopeEncryption                                        # <1>
+  config:
+    kms: <KMS service name>                                       # <2>
+    kmsConfig:                                                    # <3>
+      ..:
+    selector: <KEK selector service name>                         # <4>
+    selectorConfig:                                               # <5>
+      ..:
+----
+<1> The name of the filter. This must be `EnvelopeEncryption`.
+<2> The KMS service name.
+<3> Object providing configuration understood by KMS provider.
+<4> The KEK selector service name.
+<5> Object providing configuration understood by key selector.
+
+==== KMS Service configuration
+
+In order to configure the KMS Service, follow these KMS provider specific steps.
+
+include::hashicorp-vault/service.adoc[leveloffset=3]
+
+==== KEK selector configuration
+
+The role of the KEK selector is to map from the topic name to key name.  The filter looks up the resulting
+key name in the KMS.
+
+NOTE: If the filter is unable to find the key in the KMS, the filter will pass through the
+records belonging to that topic in the produce request without encrypting them.
```

Comment:
I still think this represents a bug. As it allows silent failure of the selector policy.

Reply:
We already have the issue https://github.com/kroxylicious/kroxylicious/issues/811